mirrord/agent/src/outgoing.rs

```diff
@@ -160,89 +78,267 @@ impl TcpOutgoingApi {
+                let daemon_connect = time::timeout(
+                    Self::CONNECT_TIMEOUT,
+                    SocketStream::connect(remote_address.clone(), self.pid),
+                )
+                .await
+                .unwrap_or_else(|_elapsed| {
+                    tracing::warn!(
+                        %remote_address,
+                        connect_timeout_ms = Self::CONNECT_TIMEOUT.as_millis(),
+                        "Connect attempt timed out."
+                    );
+
+                    Err(ResponseError::Remote(RemoteError::ConnectTimedOut(
+                        remote_address.clone(),
+                    )))
+                })
+                .and_then(|remote_stream| {
+                    let agent_address = remote_stream.local_addr()?;
+                    let connection_id = self.next_connection_id;
+                    self.next_connection_id += 1;
+
+                    let (read_half, write_half) = io::split(remote_stream);
+                    self.writers.insert(connection_id, write_half);
+                    self.readers.insert(
+                        connection_id,
+                        ReaderStream::with_capacity(read_half, Self::READ_BUFFER_SIZE),
+                    );
+
+                    Ok(DaemonConnect {
+                        connection_id,
+                        remote_address,
+                        local_address: agent_address,
+                    })
+                });
+
+                self.daemon_tx
+                    .send(DaemonTcpOutgoing::Connect(daemon_connect))
+                    .await?
+            }
 
-    /// Receives a `TcpOutgoingResponse` from the `interceptor_task`.
-    pub(crate) async fn daemon_message(&mut self) -> Result<DaemonTcpOutgoing> {
-        match self.daemon_rx.recv().await {
-            Some(msg) => Ok(msg),
-            None => Err(self.task_status.unwrap_err().await),
+            LayerTcpOutgoing::Write(LayerWrite {
+                connection_id,
+                bytes,
+            }) => {
+                let write_result = match self.writers.get_mut(&connection_id) {
+                    Some(writer) if bytes.is_empty() => {
+                        tracing::trace!(
+                            connection_id,
+                            "Received 0-sized write from layer, shutting down peer connection."
+                        );
+
+                        writer.shutdown().await.map_err(ResponseError::from)
+                    }
+
+                    Some(writer) => writer.write_all(&bytes).await.map_err(ResponseError::from),
+
+                    None => Err(ResponseError::NotFound(connection_id)),
+                };
+
+                match write_result {
+                    Ok(()) if bytes.is_empty() => {
+                        self.writers.remove(&connection_id);
+
+                        if !self.readers.contains_key(&connection_id) {
```

Comment:
I couldn't find when we ever remove a connection's reader before we remove the writer, so maybe document what this case is and when does it happen.